lib/smack/MemorySafetyChecker.cpp

```diff
@@ -3,30 +3,63 @@
 //
 #include "smack/MemorySafetyChecker.h"
 #include "smack/Naming.h"
+#include "smack/SmackOptions.h"
 #include "llvm/IR/Module.h"
 #include "llvm/IR/Instructions.h"
 #include "llvm/IR/InstIterator.h"
 #include "llvm/IR/Constants.h"
 #include "llvm/Support/Debug.h"
 #include "llvm/IR/ValueSymbolTable.h"
+#include "llvm/IR/IntrinsicInst.h"
 
 namespace smack {
 
 using namespace llvm;
 
+void insertMemoryLeakStatement(Function& F, Module& m) {
+  Function* memoryLeakCheckFunction = m.getFunction(Naming::MEMORY_LEAK_CHECK_FUNCTION);
+  for(inst_iterator I = inst_begin(F), E = inst_end(F); I != E; ++I) {
+    if(isa<ReturnInst>(&*I)) {
+      CallInst::Create(memoryLeakCheckFunction, "", &*I);
+    }
+  }
+}
+
 bool MemorySafetyChecker::runOnModule(Module& m) {
   DataLayout* dataLayout = new DataLayout(&m);
   Function* memorySafetyFunction = m.getFunction(Naming::MEMORY_SAFETY_FUNCTION);
   assert(memorySafetyFunction != NULL && "Couldn't find memory safety function");
   for (auto& F : m) {
     if (!Naming::isSmackName(F.getName())) {
+      if (SmackOptions::isEntryPoint(F.getName())) {
+        insertMemoryLeakStatement(F, m);
+      }
       for (inst_iterator I = inst_begin(F), E = inst_end(F); I != E; ++I) {
         Value* pointer = NULL;
         if (LoadInst* li = dyn_cast<LoadInst>(&*I)) {
           pointer = li->getPointerOperand();
         } else if (StoreInst* si = dyn_cast<StoreInst>(&*I)) {
           pointer = si->getPointerOperand();
-        }
+        } else if (MemSetInst* memseti = dyn_cast<MemSetInst>(&*I)) {
+	    Value* dest = memseti->getArgOperand(0);
+	    Value* size = memseti->getArgOperand(2);
+	    Type* voidPtrTy = PointerType::getUnqual(IntegerType::getInt8Ty(F.getContext()));
+	    CastInst* castPtr = CastInst::Create(Instruction::BitCast, dest, voidPtrTy, "", &*I);
+	    Value* args[] = {castPtr, size};
+	    CallInst::Create(memorySafetyFunction, ArrayRef<Value*>(args, 2), "", &*I);
+        } else if (MemCpyInst* memcpyi = dyn_cast<MemCpyInst>(&*I)) {
+	    Value* dest = memcpyi->getArgOperand(0);
+	    Value* src = memcpyi->getArgOperand(1);
+	    Value* size = memcpyi->getArgOperand(2);
+	    Type* voidPtrTy = PointerType::getUnqual(IntegerType::getInt8Ty(F.getContext()));
+	    CastInst* castPtrDest = CastInst::Create(Instruction::BitCast, dest, voidPtrTy, "", &*I);
+	    CastInst* castPtrSrc = CastInst::Create(Instruction::BitCast, src, voidPtrTy, "", &*I);
+	    Value* argsDest[] = {castPtrDest, size};
+	    Value* argsSrc[] = {castPtrSrc, size};
+	    CallInst::Create(memorySafetyFunction, ArrayRef<Value*>(argsDest, 2), "", &*I);  
+	    CallInst::Create(memorySafetyFunction, ArrayRef<Value*>(argsSrc, 2), "", &*I);  
+	}
```

Comment:
Sorry for mingling in. You could probably pass `{castPtrDest, size}` as the second argument of `CallInst::Create` directly.